### PR TITLE
Support RWX volume filesystem trim

### PR DESCRIPTION
**Pull request by derekbit**
Depends on https://github.com/longhorn/longhorn-share-manager/pull/59

https://github.com/longhorn/longhorn/issues/5143

### CODE REVIEW

File: controller/share_manager_controller.go [hunk 1]
```diff
@@ -567,8 +567,10 @@ func (c *ShareManagerController) cleanupShareManagerPod(sm *longhorn.ShareManage
 // controls transitions to running, error
 func (c *ShareManagerController) syncShareManagerPod(sm *longhorn.ShareManager) (err error) {
 	defer func() {
-		if sm.Status.State == longhorn.ShareManagerStateStopping || sm.Status.State == longhorn.ShareManagerStateStopped ||
+		if sm.Status.State == longhorn.ShareManagerStateStopping ||
+			sm.Status.State == longhorn.ShareManagerStateStopped ||
 			sm.Status.State == longhorn.ShareManagerStateError {
+			sm.Status.IP = ""
```
Comment (innobead) [inline controller/share_manager_controller.go:573]: Why requiring `sm.Status.IP = ""`? Don't the other 3 state conditions cover all cases to clean up the SM pod?

File: engineapi/share_manager.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	grpcClient *smclient.ShareManagerClient
+}
+
+func NewShareManagerClient(sm *longhorn.ShareManager) (*ShareManagerClient, error) {
+	if sm.Status.State != longhorn.ShareManagerStateRunning {
+		return nil, fmt.Errorf("invalid Share Manager %v, state: %v", sm.Name, sm.Status.State)
+	}
+
+	return &ShareManagerClient{
+		ip:         sm.Status.IP,
```
Comment (innobead) [inline engineapi/share_manager.go:22]: This is control path, so nothing related to storage network case.

File: manager/volume.go [hunk 3]
```diff
@@ -677,7 +678,27 @@ func (m *VolumeManager) TrimFilesystem(name string) (v *longhorn.Volume, err err
...
+	sm, err := m.ds.GetShareManager(volumeName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get share manager for trimming volume %v", volumeName)
+	}
+	client, err := engineapi.NewShareManagerClient(sm)
+	if err != nil {
+		return errors.Wrapf(err, "failed to launch gRPC client for share manager before trimming volume %v", volumeName)
+	}
+	err = client.FilesystemTrim()
+	return err
```
Comment (innobead) [inline manager/volume.go:701]:
```go
	return client.FilesystemTrim()
```

**Review by innobead — CHANGES_REQUESTED** [review]
In general, LGTM. Just few comments needed to resolve.

File: engineapi/share_manager.go (hunk 2 above)
Comment (derekbit) [reply engineapi/share_manager.go]: Addressed. Thank you.

File: manager/volume.go (hunk 3 above)
Comment (derekbit) [reply manager/volume.go]: Updated. Thank you.

File: controller/share_manager_controller.go (hunk 1 above)
Comment (derekbit) [reply controller/share_manager_controller.go]: It is not necessary now. Removed. Thank you.

**Comment by derekbit** [comment]
Dependencies are updated. cc @innobead

File: manager/volume.go [hunk 4]
```diff
@@ -680,7 +681,31 @@ func (m *VolumeManager) TrimFilesystem(name string) (v *longhorn.Volume, err err
...
 	}
 
-	return v, util.TrimFilesystem(name, v.Spec.Encrypted)
+	if v.Spec.AccessMode == longhorn.AccessModeReadWriteMany {
+		return v, m.trimRWXVolumeFilesystem(name, v.Spec.Encrypted)
+	}
+	return v, m.trimNonRWXVolumeFilesystem(name, v.Spec.Encrypted)
+}
+
+func (m *VolumeManager) trimNonRWXVolumeFilesystem(volumeName string, isEncryptedDevice bool) error {
```
Comment (innobead) [inline manager/volume.go:690]: `encrypted bool` instead.

File: engineapi/share_manager.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+		return nil, errors.Wrapf(err, "failed to create Share Manager client for %v", sm.Name)
+	}
+
+	return &ShareManagerClient{
+		grpcClient: client,
+	}, nil
+}
+
+func (c *ShareManagerClient) Close() error {
+	if c == nil || c.grpcClient == nil {
```
Comment (innobead) [inline engineapi/share_manager.go:35]: `if c == nil` is needed?

File: controller/share_manager_controller.go [hunk 6]
```diff
@@ -610,10 +610,11 @@ func (c *ShareManagerController) syncShareManagerPod(sm *longhorn.ShareManager)
 	// A new pod will be recreated by the share manager controller.
 	isDown, err := c.ds.IsNodeDownOrDeleted(pod.Spec.NodeName)
 	if err != nil {
-		log.WithError(err).Warnf("cannot check IsNodeDownOrDeleted(%v) when syncShareManagerPod", pod.Spec.NodeName)
+		log.WithError(err).Warnf("Cannot check IsNodeDownOrDeleted(%v) when syncShareManagerPod", pod.Spec.NodeName)
```
[the remark this answers is not in the document]
Comment (derekbit) [reply controller/share_manager_controller.go]: Updated. Thank you.

File: engineapi/share_manager.go (hunk 5 above)
Comment (derekbit) [reply engineapi/share_manager.go]: Updated. Thank you.

File: manager/volume.go (hunk 4 above)
Comment (derekbit) [reply manager/volume.go]: Updated. Thank you.

File: manager/volume.go [hunk 7]
```diff
@@ -680,7 +681,31 @@ func (m *VolumeManager) TrimFilesystem(name string) (v *longhorn.Volume, err err
...
+	return util.TrimFilesystem(volumeName, isEncryptedDevice)
+}
+
+func (m *VolumeManager) trimRWXVolumeFilesystem(volumeName string, isEncryptedDevice bool) error {
```
[the remark this answers is not in the document]
Comment (derekbit) [reply manager/volume.go:694]: Updated. Thank you.

File: manager/volume.go [hunk 8]
```diff
@@ -680,7 +681,31 @@ func (m *VolumeManager) TrimFilesystem(name string) (v *longhorn.Volume, err err
...
+		return v, m.trimRWXVolumeFilesystem(name, v.Spec.Encrypted)
+	}
+	return v, m.trimNonRWXVolumeFilesystem(name, v.Spec.Encrypted)
+}
+
+func (m *VolumeManager) trimNonRWXVolumeFilesystem(volumeName string, encryptedDevice bool) error {
+	return util.TrimFilesystem(volumeName, encryptedDevice)
+}
+
+func (m *VolumeManager) trimRWXVolumeFilesystem(volumeName string, isEncryptedDevice bool) error {
```
Comment (innobead) [inline manager/volume.go:694]: This is not changed yet :slightly_smiling_face: 

`encryptedDevice`

**Review by innobead — CHANGES_REQUESTED** [review]

File: manager/volume.go (hunk 8 above)
Comment (derekbit) [reply manager/volume.go]: Updated. Thank you.

**Review by innobead — APPROVED** [review]

**Pull request merged**